File: app/src/main/java/com/rhm/pwn/model/URLCheckTask.java
package com.rhm.pwn.model;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.SystemClock;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.text.TextUtils;
import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Selector;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.annotations.NonNull;
import io.reactivex.schedulers.Schedulers;
import java8.util.stream.StreamSupport;
import okhttp3.Headers;
import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.http.Header;

import com.rhm.pwn.PWNApp;
import com.rhm.pwn.R;
import com.rhm.pwn.home.PWNHomeActivity;
import com.rhm.pwn.network.PWNRetroFitConnector;
import com.rhm.pwn.utils.PWNUtils;

/**
 * Created by sambo on 9/3/2017.
 */

public class URLCheckTask {
    volatile static int COUNT = 0;
    volatile static int SKIPPED = 0;
    static int TOTAL = 0;

    public static final int MESSAGE_LEN = 768;
    public static final int LONG_MESSAGE_LEN = 768;


    public static boolean doesURLCheckRequireUpdate(URLCheck urlc, long curElapsedRealTime) {
        if (!TextUtils.isEmpty(urlc.getBaseUrl()) && (curElapsedRealTime - urlc.getLastElapsedRealtime()) > urlc.getCheckInterval()*1000) {
            PWNLog.log(URLCheckTask.class.getName(), "doesURLCheckRequireUpdate is TRUE: curElapsedRealTime="+curElapsedRealTime+" - urlc.getLastElapsedRealtime()="+urlc.getLastElapsedRealtime()+" > urlc.getCheckInterval()="+urlc.getCheckInterval());
            return true;
        }
        PWNLog.log(URLCheckTask.class.getName(), "doesURLCheckRequireUpdate is FALSE");
        return false;
    }

    public static long checkAll(List<URLCheck> urlChecks, Context appContext) {
        COUNT = 0;
        TOTAL = 0;
        PWNLog.log(URLCheckTask.class.getName(), "Start check task");
        long curElapsedTime = SystemClock.elapsedRealtime();
        long minIntervalRequested = Long.MAX_VALUE;
        int num = 0;
        for (URLCheck urlCheck : urlChecks) {
            minIntervalRequested = Math.min(minIntervalRequested, urlCheck.getCheckInterval());
            PWNLog.log(URLCheckTask.class.getName(), "Update check for #" + num + ", " + urlCheck.getDisplayTitle());
            if (URLCheckTask.doesURLCheckRequireUpdate(urlCheck, curElapsedTime)) {
                TOTAL++;
                PWNLog.log(URLCheckTask.class.getName(), "Scheduling check for #" + num);
                URLCheckTask.handleURLCheckAction(urlCheck, appContext);
            } else {
                PWNLog.log(URLCheckTask.class.getName(), "Skipping check for #" + num);
            }
            num++;
        }

        if (urlChecks.size() > 0) {
            return minIntervalRequested;
        } else {
            return 0;
        }
    }

    public static void checkForNotification(Context appContext) {
        PWNLog.log(URLCheckTask.class.getName(), "Checking if notifications are needed");

        Completable.fromAction(() -> {
            List<URLCheck> list = PWNDatabase.getInstance(appContext).urlCheckDao().getAll();
            List<URLCheck> updated = new ArrayList<>();
            StreamSupport.stream(list)
                    .filter(urlCheck -> urlCheck.isHasBeenUpdated() && !urlCheck.isUpdateShown())
                    .forEach(urlCheck -> {
                        PWNLog.log(URLCheckTask.class.getName(), "Found url updated at: " + urlCheck.getDisplayTitle());
                        updated.add(urlCheck);
                        urlCheck.setUpdateShown(true);
                        //save that the tasks have been notified to the user
                        PWNDatabase.getInstance(appContext).urlCheckDao().update(urlCheck);
                    });
            if (updated != null && updated.size() > 0) {
                PWNLog.log(URLCheckTask.class.getName(), "Some urls updated, prepare to show notifications for #" + updated.size());
                //Set the notifications
                String shortMessage = buildShortNotificationMessage(updated);
                String longMessage = buildLongNotificationMessage(updated);
                //create notification
                if (PWNUtils.isAppIsInBackground(appContext)) {
                    PWNLog.log(URLCheckTask.class.getName(), "App in the background, showing notifications");
                    createNotifications(appContext, updated);
                } else {
                    PWNLog.log(URLCheckTask.class.getName(), "App in the foreground, suppressing notifications");
                }
                //create system badge
            } else {

            }
        }).subscribeOn(Schedulers.io())
                .subscribe();
    }

    private static void checkIfAllDone(Context appContext) {
        PWNLog.log(URLCheckTask.class.getName(), "Checking if complete");

        if (COUNT >= TOTAL) {
            PWNLog.log(URLCheckTask.class.getName(), "Complete, broadcasting results update");
            URLCheckJobCompletedNotifier.getNotifier().update();
            URLCheckChangeNotifier.getNotifier().update(false);
            checkForNotification(appContext);
        } else {
            PWNLog.log(URLCheckTask.class.getName(), "Not yet complete");
        }
    }


    public static void handleURLCheckAction(URLCheck urlc, Context appContext) {
        PWNLog.log(URLCheckTask.class.getName(), "Beginning check for " + urlc.getDisplayTitle());

        if (!urlc.isUrlValid()) {
            PWNLog.log(URLCheckTask.class.getName(), "Skipping, URL not valid");
            return;
        }
        Retrofit retrofit = PWNRetroFitConnector.getInstance(urlc.getBaseUrl());
        PWNRetroFitConnector.GetPage getPageService = retrofit.create(PWNRetroFitConnector.GetPage.class);
        getPageService.GetPageAsString(urlc.getUrl()).enqueue(new Callback<String>() {
            @Override
            public void onResponse(Call<String> call, Response<String> response) {
                PWNLog.log(URLCheckTask.class.getName(), "Got response");
                PWNLog.log(URLCheckTask.class.getName(), "Headers:");
                PWNLog.log(URLCheckTask.class.getName(), response.headers().toString());
                Document doc = Jsoup.parse(response.body().toString());
                PWNLog.log(URLCheckTask.class.getName(), "JSON Parsed");
                Element tags = null;
                try {
                    PWNLog.log(URLCheckTask.class.getName(), "Searching for CSS");
                    tags = doc.select(urlc.getCssSelectorToInspect()).first();
                    if (tags != null) {
                        PWNLog.log(URLCheckTask.class.getName(), "CSS item found");
                    } else {
                        PWNLog.log(URLCheckTask.class.getName(), "CSS item NOT found");
                    }
                    PWNLog.log(URLCheckTask.class.getName(), "Getting page title");
                    String title = doc.select("title").first().text();
                    if (!TextUtils.isEmpty(title)) {
                        urlc.setTitle(title);
                        PWNLog.log(URLCheckTask.class.getName(), "Found page title: " + title);
                    } else {
                        PWNLog.log(URLCheckTask.class.getName(), "Page title NOT found");
                    }
                } catch (Selector.SelectorParseException e) {
                    PWNLog.log(URLCheckTask.class.getName(), "Error parsing selector\r\n" + e.getMessage(), "E");
                    urlc.setLastRunCode(URLCheck.CODE_RUN_FAILURE);
                    urlc.setLastRunMessage("Could not parse CSS selector. Please correct and retry.");
                } catch (IllegalArgumentException e) {
                    PWNLog.log(URLCheckTask.class.getName(), "Empty or invalid CSS\r\n" + e.getMessage(), "E");
                    urlc.setLastRunCode(URLCheck.CODE_RUN_FAILURE);
                    urlc.setLastRunMessage("CSS selector must not be empty. Please correct and retry.");
                }
                urlc.setLastChecked(Calendar.getInstance().getTime().toString());
                if (tags != null) {
                    PWNLog.log(URLCheckTask.class.getName(), "CSS result tag was ok");
                    //the content has changed from the last time
                    if (!urlc.lastValue.equals(tags.text())) {
                        urlc.setHasBeenUpdated(true);
                        urlc.setUpdateShown(false);
                    }
                    urlc.setLastValue(tags.text());
                    urlc.setLastRunCode(URLCheck.CODE_RUN_SUCCESSFUL);
                    urlc.setLastRunMessage("Last run successful");
                    urlc.setLastElapsedRealtime(SystemClock.elapsedRealtime());
                    Log.d("SAMB", URLCheckTask.class.getName() + " - Successfully completed retrieval URLCheck for: " + urlc.getUrl());
                } else {
                    //Try to use page headers for 'Last-Modified': 'Sat, 21 Apr 2018 16:35:49 GMT'
                    if (!TextUtils.isEmpty(getLastModifiedDate(response.headers()))) {
                        PWNLog.log(URLCheckTask.class.getName(), "CSS result not available, checking via Last-Modified header");
                        //the content has changed from the last time
                        if (urlc.lastValue == null || !urlc.lastValue.equals(getLastModifiedDate(response.headers()))) {
                            urlc.setHasBeenUpdated(true);
                            urlc.setUpdateShown(false);
                        }
                        urlc.setLastValue(getLastModifiedDate(response.headers()));
                        urlc.setLastRunCode(URLCheck.CODE_RUN_SUCCESSFUL);
                        urlc.setLastRunMessage("Last run successful");
                        urlc.setLastElapsedRealtime(SystemClock.elapsedRealtime());
                        Log.d("SAMB", URLCheckTask.class.getName() + " - Successfully completed retrieval URLCheck for: " + urlc.getUrl());
                    } else {
                        PWNLog.log(URLCheckTask.class.getName(), "CSS result tag was NULL");
                        urlc.setLastRunCode(URLCheck.CODE_RUN_FAILURE);
                        urlc.setLastRunMessage("Could not find the section of the page to search for. Check that CSS selector still exists for the page\n" + urlc.getCssSelectorToInspect());
                        Log.e("SAMB", URLCheckTask.class.getName() + " - Failed css retrieval URLCheck for: " + urlc.getUrl());
                    }
                }
                //save urlc to db
                Completable.fromAction(() -> {
                    PWNDatabase.getInstance(appContext).urlCheckDao().update(urlc);
                    PWNLog.log(URLCheckTask.class.getName(), "Results saved to DB");
                    COUNT++;
                    //Notify observers an update is ready
                    checkIfAllDone(appContext);
                }).subscribeOn(Schedulers.io())
                        .subscribe();
                if (tags != null) {
                    Log.d("SAMB", "Response was: " + tags.text());
                }
            }

            @Override
            public void onFailure(Call<String> call, Throwable t) {
                PWNLog.log(URLCheckTask.class.getName(), "Failure for " + urlc.getDisplayTitle() + " of \r\n" + t.getMessage());
                Log.e("SAMB", "Error Occurred", t);
                urlc.setLastChecked(Calendar.getInstance().getTime().toString());
                urlc.setLastRunCode(URLCheck.CODE_RUN_FAILURE);
                urlc.setLastRunMessage(t.getMessage());
                if (!urlc.isHTTPS()) {
                    urlc.setLastRunMessage(t.getMessage() + "\r\nTry setting URL to be https://");
                }
                //save urlc to db
                Completable.fromAction(() -> {
                    PWNDatabase.getInstance(appContext).urlCheckDao().update(urlc);
                    COUNT++;
                    Log.e("SAMB", URLCheckTask.class.getName() + " - Failed retrieval URLCheck for: " + urlc.getUrl(), t);
                    checkIfAllDone(appContext);
                }).subscribeOn(Schedulers.io())
                        .subscribe();

            }
        });
    }

    private static String getLastModifiedDate(Headers headers) {
        return headers.get("Last-Modified");
    }

    public static void createNotifications(Context appContext, List<URLCheck> updatedItems) {
        for (int index = 0; index < updatedItems.size(); index++) {
            URLCheck urlc = updatedItems.get(index);
            NotificationCompat.Builder mBuilder =
                    new NotificationCompat.Builder(appContext, PWNApp.CHANNEL_ID)
                            .setSmallIcon(R.drawable.ic_stat_name)
                            .setContentTitle(urlc.title)
                            .setContentText(urlc.getDisplayBody())
                            .setAutoCancel(true);
            mBuilder.setStyle(new NotificationCompat.BigTextStyle().bigText(buildLongNotificationMessage(urlc)));
            mBuilder.setContentText(urlc.getDisplayBody());
            // Creates an explicit intent for an Activity in your app
            Intent resultIntent = new Intent(appContext, PWNHomeActivity.class);
            resultIntent.putExtra(URLCheck.class.getName(), urlc.getId());
            resultIntent.putExtra(URLCheck.URL, urlc.getUrl());
            TaskStackBuilder stackBuilder = TaskStackBuilder.create(appContext);
            // Adds the back stack for the Intent (but not the Intent itself)
            stackBuilder.addParentStack(PWNHomeActivity.class);
            stackBuilder.addNextIntent(resultIntent);
            PendingIntent resultPendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
            mBuilder.setContentIntent(resultPendingIntent);
            NotificationManager mNotificationManager =
                    (NotificationManager) appContext.getSystemService(Context.NOTIFICATION_SERVICE);

            // mNotificationId is a unique integer your app uses to identify the
            // notification. For example, to cancel the notification, you can pass its ID
            // number to NotificationManager.cancel().
            mNotificationManager.notify(urlc.getId(), mBuilder.build());
            index++;
        }

        Completable.fromAction(() -> {
            for (URLCheck urlc : updatedItems) {
                urlc.setUpdateShown(true);
            }
            PWNLog.log(URLCheckTask.class.getName(), "Marking url checks as having been displayed to the user in DB");
            PWNDatabase.getInstance(appContext).urlCheckDao().update(updatedItems);
        }).subscribeOn(Schedulers.io())
                .subscribe();
    }

    static @NotNull String buildShortNotificationMessage(@NonNull List<URLCheck> list) {
        String message = getShortUpdateText(list.get(0).getLastValue())+"";
        return message.trim();
    }

    static @NotNull String buildLongNotificationMessage(@NonNull List<URLCheck> list) {
        String longText = "";
        for (int i = 0; i < list.size(); i++) {
            longText += getShortUpdateText(list.get(i).getDisplayTitle()) + "\n" + getLongUpdateText(list.get(i).getLastValue()) + "\n";
            if (i < list.size() - 1) {
                longText += "\n";
            }
        }
        //subtract title from this message
        String message = subtractText(longText, getTitle(list));
        return message.trim();
    }

    static @NotNull String buildLongNotificationMessage(@NonNull URLCheck item) {
        String longText = "";
        longText += getShortUpdateText(item.getDisplayTitle()) + "\n" + getLongUpdateText(item.getLastValue()) + "\n";
        //subtract title from this message
        String message = subtractText(longText, item.getTitle());
        return message.trim();
    }

    public static @NotNull String subtractText(String longText, String shortText) {
        if(longText.startsWith(shortText)) {
            return longText.substring(shortText.length());
        }
        return longText;
    }

    public static @NotNull String getShortUpdateText(@NotNull String text) {
        if (text != null && text.length() > MESSAGE_LEN) {
            return text.substring(0, MESSAGE_LEN - 1) + "…\n";
        }
        return text;
    }

    public static String getLongUpdateText(String text) {
        if (text != null && text.length() > LONG_MESSAGE_LEN) {
            return text.substring(0, LONG_MESSAGE_LEN - 1) + "…\n";
        }
        return text;
    }

    public static String getTitle(@NotNull List<URLCheck> list) {
        String title="";
        if (!list.isEmpty()) {
            title = list.get(0).title;
        }
        return title;
    }

}

File: app/src/main/java/com/rhm/pwn/model/URLCheckTask.kt
package com.rhm.pwn.model

import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.SystemClock
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder
import android.text.TextUtils
import android.util.Log
import com.rhm.pwn.PWNApp
import com.rhm.pwn.R
import com.rhm.pwn.home.PWNHomeActivity
import com.rhm.pwn.network.PWNRetroFitConnector
import com.rhm.pwn.utils.PWNUtils
import io.reactivex.Completable
import io.reactivex.schedulers.Schedulers
import okhttp3.Headers
import org.jetbrains.annotations.NotNull
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Selector
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

/**
 * Created by sambo on 9/3/2017.
 */

object URLCheckTask {
    @Volatile
    internal var COUNT = 0
    private var TOTAL = 0

    private const val MESSAGE_LEN = 768
    private const val LONG_MESSAGE_LEN = 768


    @JvmStatic
    fun doesURLCheckRequireUpdate(urlc: URLCheck, curElapsedRealTime: Long): Boolean {
        if (!TextUtils.isEmpty(urlc.baseUrl) && curElapsedRealTime - urlc.lastElapsedRealtime > urlc.checkInterval * 1000) {
            PWNLog.log(URLCheckTask::class.java.name, "doesURLCheckRequireUpdate is TRUE: curElapsedRealTime=" + curElapsedRealTime + " - urlc.getLastElapsedRealtime()=" + urlc.lastElapsedRealtime + " > urlc.getCheckInterval()=" + urlc.checkInterval)
            return true
        }
        PWNLog.log(URLCheckTask::class.java.name, "doesURLCheckRequireUpdate is FALSE")
        return false
    }

    @JvmStatic
    fun checkAll(urlChecks: List<URLCheck>, appContext: Context): Long {
        COUNT = 0
        TOTAL = 0
        PWNLog.log(URLCheckTask::class.java.name, "Start check task")
        val curElapsedTime = SystemClock.elapsedRealtime()
        var minIntervalRequested = java.lang.Long.MAX_VALUE
        for ((num, urlCheck) in urlChecks.withIndex()) {
            minIntervalRequested = Math.min(minIntervalRequested, urlCheck.checkInterval)
            PWNLog.log(URLCheckTask::class.java.name, "Update check for #" + num + ", " + urlCheck.displayTitle)
            if (URLCheckTask.doesURLCheckRequireUpdate(urlCheck, curElapsedTime)) {
                TOTAL++
                PWNLog.log(URLCheckTask::class.java.name, "Scheduling check for #$num")
                URLCheckTask.handleURLCheckAction(urlCheck, appContext)
            } else {
                PWNLog.log(URLCheckTask::class.java.name, "Skipping check for #$num")
            }
        }

        return if (urlChecks.isNotEmpty()) {
            minIntervalRequested
        } else {
            0
        }
    }

    private fun checkForNotification(appContext: Context) {
        PWNLog.log(URLCheckTask::class.java.name, "Checking if notifications are needed")

        Completable.fromAction {
            val list = PWNDatabase.getInstance(appContext).urlCheckDao().all()
            val updated = list.filter { it.hasBeenUpdated && !it.updateShown }
            updated.forEach { urlCheck ->
                PWNLog.log(URLCheckTask::class.java.name, "Found url updated at: " + urlCheck.displayTitle)
                urlCheck.updateShown = true
                //save that the tasks have been notified to the user
                PWNDatabase.getInstance(appContext).urlCheckDao().update(urlCheck)
            }
            if (updated.isNotEmpty()) {
                PWNLog.log(URLCheckTask::class.java.name, "Some urls updated, prepare to show notifications for #" + updated.size)
                //create notification
                if (PWNUtils.isAppIsInBackground(appContext)) {
                    PWNLog.log(URLCheckTask::class.java.name, "App in the background, showing notifications")
                    createNotifications(appContext, updated)
                } else {
                    PWNLog.log(URLCheckTask::class.java.name, "App in the foreground, suppressing notifications")
                }
            }
        }.subscribeOn(Schedulers.io())
                .subscribe()
    }

    private fun checkIfAllDone(appContext: Context) {
        PWNLog.log(URLCheckTask::class.java.name, "Checking if complete")

        if (COUNT >= TOTAL) {
            PWNLog.log(URLCheckTask::class.java.name, "Complete, broadcasting results update")
            URLCheckJobCompletedNotifier.notifier.update()
            URLCheckChangeNotifier.getNotifier().update(false)
            checkForNotification(appContext)
        } else {
            PWNLog.log(URLCheckTask::class.java.name, "Not yet complete")
        }
    }


    private fun handleURLCheckAction(urlc: URLCheck, appContext: Context) {
        PWNLog.log(URLCheckTask::class.java.name, "Beginning check for " + urlc.displayTitle)

        if (!urlc.urlValid) {
            PWNLog.log(URLCheckTask::class.java.name, "Skipping, URL not valid")
            return
        }
        val retrofit = PWNRetroFitConnector.getInstance(urlc.baseUrl)
        val getPageService = retrofit.create(PWNRetroFitConnector.GetPage::class.java)
        getPageService.GetPageAsString(urlc.getUrl()).enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                PWNLog.log(URLCheckTask::class.java.name, "Got response")
                PWNLog.log(URLCheckTask::class.java.name, "Headers:")
                PWNLog.log(URLCheckTask::class.java.name, response.headers().toString())
                val doc = Jsoup.parse(response.body())
                PWNLog.log(URLCheckTask::class.java.name, "JSON Parsed")
                var tags: Element? = null
                try {
                    PWNLog.log(URLCheckTask::class.java.name, "Searching for CSS")
                    tags = doc.select(urlc.cssSelectorToInspect!!).first()
                    if (tags != null) {
                        PWNLog.log(URLCheckTask::class.java.name, "CSS item found")
                    } else {
                        PWNLog.log(URLCheckTask::class.java.name, "CSS item NOT found")
                    }
                    PWNLog.log(URLCheckTask::class.java.name, "Getting page title")
                    val title = doc.select("title").first().text()
                    if (!TextUtils.isEmpty(title)) {
                        urlc.title = title
                        PWNLog.log(URLCheckTask::class.java.name, "Found page title: $title")
                    } else {
                        PWNLog.log(URLCheckTask::class.java.name, "Page title NOT found")
                    }
                } catch (e: Selector.SelectorParseException) {
                    PWNLog.log(URLCheckTask::class.java.name, "Error parsing selector\r\n" + e.message, "E")
                    urlc.lastRunCode = URLCheck.CODE_RUN_FAILURE
                    urlc.lastRunMessage = "Could not parse CSS selector. Please correct and retry."
                } catch (e: IllegalArgumentException) {
                    PWNLog.log(URLCheckTask::class.java.name, "Empty or invalid CSS\r\n" + e.message, "E")
                    urlc.lastRunCode = URLCheck.CODE_RUN_FAILURE
                    urlc.lastRunMessage = "CSS selector must not be empty. Please correct and retry."
                }

                val curDate = Calendar.getInstance().time.toString()
                urlc.lastChecked = curDate
                if (tags != null) {
                    PWNLog.log(URLCheckTask::class.java.name, "CSS result tag was ok")
                    //the content has changed from the last time
                    if (urlc.lastValue != tags.text()) {
                        urlc.hasBeenUpdated = true
                        urlc.updateShown = false
                        urlc.lastUpdated = curDate
                    }
                    urlc.lastValue = tags.text()
                    urlc.lastRunCode = URLCheck.CODE_RUN_SUCCESSFUL
                    urlc.lastRunMessage = "Last run successful"
                    urlc.lastElapsedRealtime = SystemClock.elapsedRealtime()
                    Log.d("SAMB", URLCheckTask::class.java.name + " - Successfully completed retrieval URLCheck for: " + urlc.getUrl())
                } else {
                    //Try to use page headers for 'Last-Modified': 'Sat, 21 Apr 2018 16:35:49 GMT'
                    if (!TextUtils.isEmpty(getLastModifiedDate(response.headers()))) {
                        PWNLog.log(URLCheckTask::class.java.name, "CSS result not available, checking via Last-Modified header")
                        //the content has changed from the last time
                        if (urlc.lastValue == null || urlc.lastValue != getLastModifiedDate(response.headers())) {
                            urlc.hasBeenUpdated = true
                            urlc.updateShown = false
                            urlc.lastUpdated = curDate
                        }
                        urlc.lastValue = getLastModifiedDate(response.headers())
                        urlc.lastRunCode = URLCheck.CODE_RUN_SUCCESSFUL
                        urlc.lastRunMessage = "Last run successful"
                        urlc.lastElapsedRealtime = SystemClock.elapsedRealtime()
                        Log.d("SAMB", URLCheckTask::class.java.name + " - Successfully completed retrieval URLCheck for: " + urlc.getUrl())
                    } else {
                        PWNLog.log(URLCheckTask::class.java.name, "CSS result tag was NULL")
                        urlc.lastRunCode = URLCheck.CODE_RUN_FAILURE
                        urlc.lastRunMessage = "Could not find the section of the page to search for. Check that CSS selector still exists for the page\n" + urlc.cssSelectorToInspect!!
                        Log.e("SAMB", URLCheckTask::class.java.name + " - Failed css retrieval URLCheck for: " + urlc.getUrl())
                    }
                }
                //save urlc to db
                Completable.fromAction {
                    PWNDatabase.getInstance(appContext).urlCheckDao().update(urlc)
                    PWNLog.log(URLCheckTask::class.java.name, "Results saved to DB")
                    COUNT++
                    //Notify observers an update is ready
                    checkIfAllDone(appContext)
                }.subscribeOn(Schedulers.io())
                        .subscribe()
                if (tags != null) {
                    Log.d("SAMB", "Response was: " + tags.text())
                }
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                PWNLog.log(URLCheckTask::class.java.name, "Failure for " + urlc.displayTitle + " of \r\n" + t.message)
                Log.e("SAMB", "Error Occurred", t)
                urlc.lastChecked = Calendar.getInstance().time.toString()
                urlc.lastRunCode = URLCheck.CODE_RUN_FAILURE
                urlc.lastRunMessage = t.message
                if (!urlc.isHTTPS) {
                    urlc.lastRunMessage = t.message + "\r\nTry setting URL to be https://"
                }
                //save urlc to db
                Completable.fromAction {
                    PWNDatabase.getInstance(appContext).urlCheckDao().update(urlc)
                    COUNT++
                    Log.e("SAMB", URLCheckTask::class.java.name + " - Failed retrieval URLCheck for: " + urlc.getUrl(), t)
                    checkIfAllDone(appContext)
                }.subscribeOn(Schedulers.io())
                        .subscribe()

            }
        })
    }

    private fun getLastModifiedDate(headers: Headers): String? {
        return headers.get("Last-Modified")
    }

    private fun createNotifications(appContext: Context, updatedItems: List<URLCheck>) {
        var index = 0
        while (index < updatedItems.size) {
            val urlc = updatedItems[index]
            val mBuilder = NotificationCompat.Builder(appContext, PWNApp.CHANNEL_ID)
                    .setSmallIcon(R.drawable.ic_stat_name)
                    .setContentTitle(urlc.title)
                    .setContentText(urlc.displayBody)
                    .setAutoCancel(true)
            mBuilder.setStyle(NotificationCompat.BigTextStyle().bigText(buildLongNotificationMessage(urlc)))
            mBuilder.setContentText(urlc.displayBody)
            // Creates an explicit intent for an Activity in your app
            val resultIntent = Intent(appContext, PWNHomeActivity::class.java)
            resultIntent.putExtra(URLCheck.CLASSNAME, urlc.id)
            resultIntent.putExtra(URLCheck.URL, urlc.getUrl())
            val stackBuilder = TaskStackBuilder.create(appContext)
            // Adds the back stack for the Intent (but not the Intent itself)
            stackBuilder.addParentStack(PWNHomeActivity::class.java)
            stackBuilder.addNextIntent(resultIntent)
            val resultPendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT)
            mBuilder.setContentIntent(resultPendingIntent)
            val mNotificationManager = appContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

            // mNotificationId is a unique integer your app uses to identify the
            // notification. For example, to cancel the notification, you can pass its ID
            // number to NotificationManager.cancel().
            mNotificationManager.notify(urlc.id, mBuilder.build())
            index++
            index++
        }

        Completable.fromAction {
            for (urlc in updatedItems) {
                urlc.updateShown = true
            }
            PWNLog.log(URLCheckTask::class.java.name, "Marking url checks as having been displayed to the user in DB")
            PWNDatabase.getInstance(appContext).urlCheckDao().update(updatedItems)
        }.subscribeOn(Schedulers.io())
                .subscribe()
    }

    @JvmStatic
    fun buildLongNotificationMessage(@NotNull item: URLCheck): String {
        var longText = ""
        longText += getShortUpdateText(item.displayTitle) + "\n" + getLongUpdateText(item.lastValue
                ?: "") + "\n"
        //subtract title from this message
        val message = subtractText(longText, item.title ?: "")
        return message.trim { it <= ' ' }
    }

    private fun subtractText(@NotNull longText: String, @NotNull shortText: String): String {
        return if (longText.startsWith(shortText)) {
            longText.substring(shortText.length)
        } else longText
    }

    private fun getShortUpdateText(@NotNull text: String): String {
        return if (text.length > MESSAGE_LEN) {
            text.substring(0, MESSAGE_LEN - 1) + "…\n"
        } else text
    }

    private fun getLongUpdateText(@NotNull text: String): String {
        return if (text.length > LONG_MESSAGE_LEN) {
            text.substring(0, LONG_MESSAGE_LEN - 1) + "…\n"
        } else text
    }
}
